zebra-network/src/peer/connection.rs

```diff
@@ -312,25 +311,62 @@ impl Handler {
+/// The current state of the [`Connection`], consumed to execute the next step of
+/// the state machine.
 pub(super) enum State {
     /// Awaiting a client request or a peer message.
     AwaitingRequest,
     /// Awaiting a peer message we can interpret as a client request.
     AwaitingResponse {
+        /// Inner state machine for handling external responses.
         handler: Handler,
+        /// Channel used to propagate responses back to the [`Client`] in our
+        /// internal Response format.
         tx: MustUseOneshotSender<Result<Response, SharedPeerError>>,
         span: tracing::Span,
     },
 }
 
 impl State {
+    /// Execute one step of the [`Connection`] state machine event loop. This
+    /// function represents the core logic of [`Connection::run`] method and
+    /// isolates consuming the previous state and producing the next state into a
+    /// single function.
+    ///
+    /// This function's primary purpose is to provide compile time guarantees
+    /// that iterations of the run loop never leave the Connection with an
+    /// invalid `state`, by forcing all code paths to produce state transition in
+    /// order to exit the function.
     async fn step<Rx, S, Tx>(self, conn: &mut Connection<S, Tx>, peer_rx: &mut Rx) -> Transition
     where
         Rx: Stream<Item = Result<Message, SerializationError>> + Unpin,
         S: Service<Request, Response = Response, Error = BoxError>,
         S::Error: Into<BoxError>,
         Tx: Sink<Message, Error = SerializationError> + Unpin,
     {
+        // At a high level, the event loop we want is as follows: we check for
+        // any incoming messages from the remote peer, check if they should be
+        // interpreted as a response to a pending client request
+        // (Handler::process_request), and if not, interpret them as a request
+        // from the remote peer to our node
+        // (Connection::handle_message_as_request/drive_peer_request).
+        //
+        // We also need to handle those client requests in the first place
+        // (Connection::handle_client_request). The client requests are received
+        // from the corresponding `peer::Client` over a bounded channel (with
+        // bound 1, to minimize buffering), but there is no relationship between
+        // the stream of client requests and the stream of peer messages, so we
+        // cannot ignore one kind while waiting on the other. Moreover, we
+        // cannot accept a second client request while the first one is still
+        // pending.
+        //
+        // To do this, we inspect the current request state.
+        //
+        // If there is no pending request, we wait on either an incoming peer message or
+        // an incoming request, whichever comes first.
+        //
+        // If there is a pending request, we wait only on an incoming peer message, and
+        // check whether it can be interpreted as a response to the pending request.
```

Comment:
👍 

These details might be worth linking/moving to the top of the module. But that shouldn't block the PR.

Reply:
Feel free to do this change later, or commit and then admin-merge.

Reply:
I don't know, its pretty specifically tied to this code, my feeling is that it's best to keep it near the logic it is describing, since this is implementation details and not an explanation of how to interact with it

Reply:
Great that works for me